frame/system/src/lib.rs

```diff
@@ -1203,12 +1231,22 @@ impl<T: Trait + Send + Sync> CheckWeight<T> where
 		let current_weight = Module::<T>::all_extrinsics_weight();
 		let maximum_weight = T::MaximumBlockWeight::get();
 		let limit = Self::get_dispatch_limit_ratio(info.class) * maximum_weight;
-		let added_weight = info.weight.min(limit);
-		let next_weight = current_weight.saturating_add(added_weight);
-		if next_weight > limit && info.class != DispatchClass::Mandatory {
-			Err(InvalidTransaction::ExhaustsResources.into())
-		} else {
+		if info.class == DispatchClass::Mandatory {
+			// If we have a dispatch that must be included in the block, we will do everything to
+			// make sure we do not return an error here.
+			let extrinsic_weight = info.weight.saturating_add(T::ExtrinsicBaseWeight::get());
+			let next_weight = current_weight.saturating_add(extrinsic_weight);
 			Ok(next_weight)
+		} else {
+			let extrinsic_weight = info.weight.checked_add(T::ExtrinsicBaseWeight::get())
+				.ok_or(InvalidTransaction::ExhaustsResources)?;
+			let next_weight = current_weight.checked_add(extrinsic_weight)
+				.ok_or(InvalidTransaction::ExhaustsResources)?;
+			if next_weight > limit {
+				Err(InvalidTransaction::ExhaustsResources.into())
+			} else {
+				Ok(next_weight)
+			}
```

Comment:
We should be able to include an Operationnal transaction which cost less than `(1 - AvailableBlockRatio) * maximum_weight` and a Normal transaction with the max_extrinsic_weight for normal transaction no ?

But I think in the current code including a Operational transaction will add into `all_extrinsics_weight` and so this above property will fail.

Reply:
Lets use regular numbers here.

So max block weight is 1000.

There is a tx with weight 750 (75% of the max block weight, takes up all `Normal` tx space).

Now there is an operational transaction of size 250, and this should succeed to be placed in the block?

I agree. Let me check the code

Reply:
I dont think this scenario will fail, but I am happy to write a test.

Reply:
both:
* adding a operational transaction of weight 250 and then normal transaction of weight 750
* adding a operational transaction of weight 750 and then normal transaction of weight 250

should work but yes

Reply:
having an operational of 250 will make all_extrinsic_weight == 250,
then checking the normal transaction of weight 750 will fail at `next_weight <= limit` as `1000< 750`

Reply:
@thiolliere I think the old code would also fail with 250 + 750 added in that order. I always considered it a thing that "a normal tx in the queue cannot push the block past 75%.

Indeed we could write the logic to be a bit more clever in this way, but that would be another behavior change i believe.

I have added a test for 750 + 250 you can see here:
https://github.com/paritytech/substrate/pull/5722/commits/09b27ecdf234857933769a79dc281706808ad20a

Reply:
I can confirm that the old logic would not be able to handle 250 + 750 either, and it actually becomes a hard problem to solve since we do not know what kinds of operations have contributed to the total weight.

I think that this is practically the best we can do without introducing additional storage items.

Reply:
The solution would be to have an `all_extrinsics_weight_normal` storage item and a `all_extrinsics_weight_operational` storage item, and to increment each one respectively, and check their sum does not push past the 100% block threshold, and that `all_extrinsics_weight_normal` is always less than the available block ratio.

So one additional storage item. I will consult with @kianenigma to see if this was already decided against

Reply:
we could store both total_weight and operational_mandatory_weight into a single storage item maybe. But otherwise ok we can say that normal transaction can only consume some weight until some point. maybe adding a few comment in the doc would be better.

Reply:
yeah a tuple or struct would work. If it is intentional from Kian, we will keep it the same, else adding the extra functionality would be simple enough.

Reply:
I initially didn't intend to see the weights types separated in any way. My only rules are: Operational can use all the weight. Normal can use 75% of it. i.e. each class of transaction can only fill up to x percent of the total weight. Normal can fill up to 75. Operational and Mandatory can fill the entire 100%. I see it it as a block being filled as such and the limit stopping new transactions filling the blocks. 
 
In your examples above, I would **not** assume that if there is a `250 operational` one, a `750 normal` can fit. A `500 normal` can fit at most, since all the normal transactions can _at most_ consume 75%. 

I think both approaches work fine and I don't have a strong opinion in favour of keeping it the way it is, except that it is simple the understand and code. So I recommend you to also not change it into a tuple approach unless if you have a strong opinion that the current one is wrong. Maintaining a tuple would be hard since it is hardcoding the types into elements, and imagining that we would add more transaction types, this can become a mess to maintain.

Reply:
The main benefit of your approach from a theoretical perspective is that the current system has starvation. If there are lots of operational transactions included at the beginning of the block, and fill 80% of the block, then the other 20% cannot be filled by normal ones and they essentially starve. 

Though I don't think this is bad in every case.

Reply:
ok we can keep as it is, the doc is actually fine it can just be understood in both ways.

Reply:
We will keep it as is for now, and we can follow up. Ill open an issue

Reply:
https://github.com/paritytech/substrate/issues/5758